File: Problem#3/euler003.cpp
#include <bits/stdc++.h>

using namespace std;

int main(){
    long long number = 600851475143;
    long long largestPrimeFactor = 1;
    for(long long primeFactor = 2; primeFactor * primeFactor <= number; ++primeFactor){
        if(number % primeFactor) continue;
        largestPrimeFactor = primeFactor;
        while(number % primeFactor == 0) number /= primeFactor;
    }

    largestPrimeFactor = max(number, largestPrimeFactor);
    cout << largestPrimeFactor << '\n';

    return 0;
}

File: Problem#3/euler003.java
public class euler003 {
   public static void main(String args[]){
       final PrimeFactorization primeFactorization = new PrimeFactorization();
       System.out.println(primeFactorization.solve());
       return;
   } 
}

class PrimeFactorization implements EulerSolution {
    public String solve(){
        Long number = 600851475143L;
        for(Long primeFactor = 2L; primeFactor * primeFactor <= number; ++primeFactor){
            if(number % primeFactor > 0) continue;
            while((number % primeFactor == 0) && number != primeFactor){
                number /= primeFactor;
            }
        }

        return String.valueOf(number);
    }
}

interface EulerSolution {
    public String solve();
}